WIP: Tutorial.

amanagr opened the pull request:
This PR brings following changes:
* Shows a Tutorial at startup.
* Allows user to change settings before loading UI.

amanagr committed:
views: Add `LoadingView` to show at startup.

This is just an extension of urwid.Text which we
constantly update until model and view are completely
loaded.

amanagr committed:
tutorial: Add tutorial to be shown at startup.

This contains basic commands and instructions
a user must know to use the app.

amanagr committed:
run: Pass zuliprc_path and additional theme settings to Controller.

These will be later used to show current config on loading screen.

amanagr committed:
core: Extract loop initialization into a separate function.

This will allow up to set custom loop properties without
running it.

amanagr committed:
core: Load `model` and `view` in asych.

Move model and view in `init_model_view` which
is an `asynch` function. This will allow us to initialize
controller with any view we want and
later change it with main view (since it is being
loaded independently).

amanagr committed:
core: Add method to update zuliprc from controller.

This can used to update zuliprc while the app is running
and set custom setting for user via UI.

amanagr committed:
core: Use show_main_view to show custom loading text.

This is used to show Tutorial while the view is loading.

amanagr committed:
run: Remove starting output form run.py

The starting text will be shown via Controller now.
The tests will also will be moved there in a later commit.

amanagr committed:
core: Remove Show loading indicator.

This has already been moved inside `show_main_view`.

amanagr commented:
I think this is mostly ready feature wise, only things remaining are:
* Adding tests
* Improve Tutorial
* Restructure commits.

I would appreciate any feedback on this! (code wise too)

shreyamalviya commented on zulipterminal/config/tutorial.py:
```diff
@@ -0,0 +1,24 @@
+
+For reading
+
+- Use the left sidebar (or the `n` key) to review messages thread-by-thread.
```
Maybe "review unread messages topic-by-topic", instead? Also, we can navigate through unread streams using the left sidebar, but unread topics using `n`, so maybe separate those two?

shreyamalviya commented on zulipterminal/config/tutorial.py:
```diff
@@ -0,0 +1,24 @@
+
+- Use the left sidebar (or the `n` key) to review messages thread-by-thread.
+- Use the down arrow key (`↓`/`j`) to move the cursor to the next message.
+- Use the End key to mark all messages in the current view as read.
```
`end` instead of End?

shreyamalviya commented on zulipterminal/config/tutorial.py:
```diff
@@ -0,0 +1,24 @@
+- You can type your message in the bottom big box and send it using `ctrl d` or `alt enter`.
+
+\n\n
+"""
```
I think we should mention at the end that pressing `?` would open up the help menu for more.

sumanthvrao commented:
@amanagr I had a go with this today. It's great overall! :+1:  I had few topics to address/request. Futher discussion can take place on these of course :
* In this PR, adding `h` to skip tutorial hence-forth is a nice touch. But suppose, I want to again enable it on startup then the only way I can do that is by changing the zulilprc `tutorial` flag manually. So, perphaps, like we do with the `theme` setting, we can add a flag (`--enable-tutorial = True`) so that the arg-parser can use it to override the setting in zuliprc 
* The only thing I felt odd about the startuppage was the way text was aligned. Maybe we can stick to left/right indentation rather than centre. Not sure how this will look though.

I skimmed through the code in each commit but haven't yet done an indepth code-wise review.

amanagr commented:
@sumanthvrao I thinking of doing a followup PR after this where we can toggle any setting in the `LoadingView`. User can just press `a` to switch from `autohide` to `no_autohide` and so on...

amanagr committed:
core: Handle exceptions in init_model_view gracefully.

Use watch_pipe to move exceptions from
`init_model_view` thread to main thread. This
allow to handle errors gracefully in the
main loop itself.

amanagr committed:
core: Add tutorial config and related settings.

amanagr commented:
@neiljp this is already getting hard to manage, can we merge some initial commits?

amanagr committed:
core: Display notify config in tutorial.

amanagr committed:
LoadingView: Allow setting autohide behaviour from Tutorial Page.

amanagr committed:
tutorial: Allow changing UI settings after model loading is completed.

amanagr committed:
Tutorial: Allow changing theme from loading view.

amanagr commented:
@neiljp @sumanthvrao I would love to have this PR reviewed before proceeding further with writing tests. This brings a huge change to how zt loads up and I would welcome suggestions on if we would do it better in a different way.

neiljp commented on zulipterminal/config/tutorial.py:
```diff
@@ -0,0 +1,18 @@
+# Tutorial to display at start
+TUTORIAL = [
```
To echo @shreyamalviya's point previously, after this text (independently of showing the tutorial text), I think it would be good to show a '"Remember you can use '?' to explore the available shortcut/hot keys" message or similar. That would act as a 'look <here> for more keys than we just introduced' as well as a 'remember you can get help <here>' if the tutorial is not enabled.

neiljp commented on zulipterminal/config/tutorial.py:
```diff
@@ -0,0 +1,18 @@
+    ('unread', '\nFor Reading\n'),
+    '- Use the left sidebar (or the `n` key) to review messages '
+    'thread-by-thread.\n',
+    '- Use the End key to mark all messages in the current view as read.',
```
As Shreya said, this key needs a highlight/tag.

neiljp commented on zulipterminal/cli/run.py:
```diff
@@ -200,7 +200,8 @@ def main(options: Optional[List[str]]=None) -> None:
             theme_to_use = zterm['theme']
         available_themes = all_themes()
-        if theme_to_use[0] not in available_themes:
+        theme = theme_to_use[0]
```
Based on other renaming here, `theme_name`?

This first commit feels like a combination of 'preparing for other commits' elements, the most obvious of which is working with just the theme names in run.py, and only accessing the theme itself in core.py. The other element is setting the zuliprc_path in core.py, which perhaps should go with another commit. Then you can describe the intent of this first commit in terms of just that theme_name vs theme change.

neiljp commented on zulipterminal/cli/run.py:
```diff
@@ -239,7 +240,8 @@ def main(options: Optional[List[str]]=None) -> None:
         Controller(zuliprc_path,
-                   THEMES[theme_to_use[0]],
+                   theme,
+                   zuliprc_path,
```
You're passing this parameter twice now?

neiljp commented on zulipterminal/core.py:
```diff
@@ -39,10 +39,7 @@ def __init__(self, config_file: str, theme_name: str,
-        self.view = View(self)
-        # Start polling for events after view is rendered.
-        self.model.poll_for_events()
+        self.init_model_view()
```
Currently `self.view` should exist before `initialize_loop` (which you just introduced), but this could cause self.view to not be set at this point.

I've not reviewed all the commits, but do we really need this to be asynch?

amanagr replied:
We don't need `self.view` to initialize_loop :) `initialize_loop` loads `LoadingView` first and then when we have `self.view` ready, we display that. Also, In the latter commits I have split the function into two parts, first I load model, then initialize loop and then load the view.

neiljp reviewed (commented):
@amanagr I see this is getting big now, so thanks for the extra ping to review this :+1: 

I've left some feedback inline, some actually regarding the implementation generally and also on some early commits, but mostly in terms of high-level thoughts as to how it behaves.

I think the concept and functionality is really great at this point, though I have some concerns about the implementation I've read so far. You seem to add code in one commit and then delete it in another; it would be clearer if you did that as a 'move' in just one commit, so we can see what is being lost vs what is being added.

amanagr commented on zulipterminal/ui_tools/views.py:
```diff
@@ -842,3 +844,53 @@ def keypress(self, size: Tuple[int, int], key: str) -> str:
+                self.controller.update_screen()
+            if key == 'h' and self.controller.wait_after_loading:
+                # Add tutorial=skip to the generated zuliprc
+                self.controller.update_zuliprc('tutorial', 'skip')
```
This is a great feature, which I think I will add as a v2 or possibly when we have less commits to keep track of in this PR. For now, users will have to remove `tutorial=skip` from the zuliprc to see this menu back. (We can add this info to the README.)

sumanthvrao commented on zulipterminal/core.py:
```diff
@@ -337,7 +316,7 @@ def spinning_cursor() -> Any:
                 for cursor in '|/-\\':
                     yield cursor
 
-        def loading_text(spinner: List[str]) -> List[Any]:
```
`List[str]` is still the type here right?

sumanthvrao reviewed (commented):
@amanagr I gave this a review. I had a go with this for most of today and it works wonderfully. 

I like how we can easily toggle settings interactively with the help of the landing page this PR takes us to. I left a few comments clarifying some small points and one which has traceback from an error I received. Most of the commits from 2nd till 7th don't work on their own. They appear to crash `zulip-term` when I run in the middle of a rebase. The left alignment of text looks better than the previous centre alignment :+1:.

As Neil mentioned in one of his comments, the way we display text in the landing page (tutorial page) can be discussed further. Maybe we can show it in the same style as the help view. ;)

amanagr commented:
@sumanthvrao thanks for your thoughts. I will look into them.
